DocGen4/Output/Pybtex.lean

```diff
@@ -0,0 +1,152 @@
+import DocGen4.Output.References
+
+/-!
+
+# bib file processor using `pybtex`
+
+This file contains functions for bib file processor using
+external program `pybtex` which is written in Python.
+
+The main function is `DocGen4.Pybtex.process`.
+
+-/
+
+open Lean DocGen4 Output
+
+namespace DocGen4
+
+namespace Pybtex
+
+private def proc (args : IO.Process.SpawnArgs) : IO Unit := do
+  let child ← IO.Process.spawn args
+  let exitCode ← child.wait
+  if exitCode != 0 then
+    throw (IO.userError s!"external command '{args.cmd}' exited with code {exitCode}")
+
+/-- Get the array of cite keys from bibtexml contents. -/
+def getCitekeys (contents : String) : Except String (Array String) :=
+  match Xml.parse contents with
+  | .ok (.Element tag _ arr) =>
+    if tag != "bibtex:file" then
+      .error s!"node is '{tag}', but 'bibtex:file' expected"
+    else
+      let arr2 : Array (Except String String) := arr.map fun node =>
+        match node with
+        | .Element (.Element tag attr _) =>
+          if tag != "bibtex:entry" then
+            .error s!"node is '{tag}', but 'bibtex:entry' expected"
+          else
+            match attr.find? "id" with
+            | .some s => .ok s
+            | .none => .error "failed to find 'id' attribute"
+        | _ => .ok ""
+      match arr2.findSome? (fun x => match x with | .error s => .some s | _ => .none) with
+      | .some s => .error s
+      | .none => .ok (arr2.map (fun x => match x with | .ok s => s | _ => unreachable!)
+        |>.filter fun s => not s.trim.isEmpty)
+  | .error err => .error err
+
+private def deleteTempFile : IO Unit := do
+  IO.FS.removeFile (basePath / "input.tmp") <|> pure ()
+  IO.FS.removeFile (basePath / "output.tmp") <|> pure ()
+
+private def getCitekeysFromTempFile : IO (Array String) := do
+  -- run `pybtex-convert`
+  proc {
+    cmd := "pybtex-convert"
+    args := #[
+      "-f", "bibtex", "-t", "bibtexml", "--preserve-case",
+      (basePath / "input.tmp").toString,
+      (basePath / "output.tmp").toString
+    ]
+  }
+  -- parse the returned XML file
```

Comment:
Would it make sense to write straight to stdout here?

Reply:
Let me study if it supports stdin/stdout.

Reply:
Unfortunately, the command line tool does not support reading from stdin and writing to stdout.